core/arch/arm/mm/core_mmu_lpae.c

```diff
@@ -339,41 +364,96 @@ static paddr_t get_nsec_ddr_max_pa(void)
 	return pa;
 }
 
-void core_init_mmu_tables(struct tee_mmap_region *mm)
+#ifdef CFG_VIRTUALIZATION
+size_t core_mmu_get_total_pages_size(void)
 {
-	uint64_t max_va = 0;
-	size_t n;
+	return ROUNDUP(sizeof(l1_xlation_table), SMALL_PAGE_SIZE) +
+		sizeof(xlat_tables) + sizeof(xlat_tables_ul1);
+}
 
-#ifdef CFG_CORE_UNMAP_CORE_AT_EL0
-	COMPILE_TIME_ASSERT(CORE_MMU_L1_TBL_OFFSET ==
-			   sizeof(l1_xlation_table) / 2);
+struct mmu_context *core_alloc_mmu_ctx(void *tables)
+{
+	struct mmu_context *ctx;
+	char *tbl = tables;
+	unsigned int asid = asid_alloc();
+
+	assert(((vaddr_t)tbl) % SMALL_PAGE_SIZE == 0);
+
+	if (!asid)
+		return NULL;
+
+	ctx = kmalloc(sizeof(*ctx));
+
+	if (!ctx)
+		return NULL;
+
+	ctx->l1_tables = (l1_xlat_tbls_t *)(void *)tbl;
+	memset(ctx->l1_tables, 0, SMALL_PAGE_SIZE);
+	tbl += ROUNDUP(sizeof(l1_xlation_table), SMALL_PAGE_SIZE);
+
+	ctx->xlat_tables = (xlat_tbl_t *)(void *)tbl;
+	memset(ctx->xlat_tables, 0, sizeof(xlat_tables));
+	tbl += sizeof(xlat_tables);
+	assert(((vaddr_t)tbl) % SMALL_PAGE_SIZE == 0);
+
+	ctx->l2_ta_tables = (xlat_tbl_t *)(void *)tbl;
+	ctx->xlat_tables_used = 0;
+	ctx->total_xlat_tables = MAX_XLAT_TABLES;
+	ctx->user_va_idx = -1;
+	ctx->asid = asid;
+
+	return ctx;
+}
+
+void core_free_mmu_ctx(struct mmu_context *ctx)
+{
+	asid_free(ctx->asid);
+	free(ctx);
+}
+
+void core_mmu_set_ctx(struct mmu_context *ctx)
+{
+	uint64_t ttbr;
+	/*
+	 * We are changing mappings for current CPU,
+	 * so make sure that we will not be rescheduled
+	 */
+	assert(thread_get_exceptions() & THREAD_EXCP_FOREIGN_INTR);
+
+	current_ctx[get_core_pos()] = ctx;
+
+	ttbr = virt_to_phys(ctx->l1_tables[0][get_core_pos()]);
+
+	write_ttbr0_el1(ttbr | ((paddr_t)ctx->asid << TTBR_ASID_SHIFT));
```

Comment:
Do we need a `dsb()` before this to make sure that the writes to the tables are visible, or is that guaranteed by other means?

Reply:
Writes are guaranteed by existing means. This commit does not add new pagetable entry manipulation. It uses existing functions like `core_mmu_map_region`.